File: 2024/Python/10.py
data = [
    [int(c) if c != "." else -1 for c in line.strip()]
    for line in open("../inputs/10.txt")
]

H = len(data)
W = len(data[0])

dydx = [(0, 1), (1, 0), (-1, 0), (0, -1)]


def valid_coord(y, x):
    return 0 <= y < H and 0 <= x < W


starts = [(y, x) for y in range(H) for x in range(W) if data[y][x] == 0]


def dfs(initial, part2=False, visited=None):
    if visited is None or part2:
        visited = set()
    if initial in visited:
        return 0
    visited.add(initial)
    y, x = initial
    val = int(data[y][x])
    if val == 9:
        return 1
    sum = 0
    for dy, dx in dydx:
        new_y, new_x = y + dy, x + dx
        if not valid_coord(new_y, new_x) or data[new_y][new_x] == ".":
            continue
        new_val = data[new_y][new_x]
        if new_val == val + 1:
            sum += dfs((new_y, new_x), part2, visited)
    return sum


scores = 0
ratings = 0

for init in starts:
    scores += dfs(init)
    ratings += dfs(init, True)

print(scores)
print(ratings)

File: 2024/Deno/10.ts
const data: number[][] = Deno.readTextFileSync("../inputs/10.txt")
    .trim()
    .split("\n")
    .map((line) => line.split("").map((c) => (c === "." ? -1 : Number(c))));

const H = data.length;
const W = data[0].length;

const dydx: [number, number][] = [[0, 1], [1, 0], [-1, 0], [0, -1]];

function validCoord(y: number, x: number): boolean {
    return 0 <= y && y < H && 0 <= x && x < W;
}

const starts: [number, number][] = [];
for (let y = 0; y < H; y++) {
    for (let x = 0; x < W; x++) {
        if (data[y][x] === 0) {
            starts.push([y, x]);
        }
    }
}

function dfs(
    initial: [number, number],
    part2 = false,
    visited: Set<string> | null = null,
): number {
    if (visited === null || part2) {
        visited = new Set();
    }
    const key = `${initial[0]},${initial[1]}`;
    if (visited.has(key)) {
        return 0;
    }
    visited.add(key);
    const [y, x] = initial;
    const val = data[y][x];
    if (val === 9) {
        return 1;
    }
    let sum = 0;
    for (const [dy, dx] of dydx) {
        const [newY, newX] = [y + dy, x + dx];
        if (!validCoord(newY, newX) || data[newY][newX] === -1) {
            continue;
        }
        const newVal = data[newY][newX];
        if (newVal === val + 1) {
            sum += dfs([newY, newX], part2, visited);
        }
    }
    return sum;
}

let scores = 0;
let ratings = 0;

for (const init of starts) {
    scores += dfs(init);
    ratings += dfs(init, true);
}

console.log(scores);
console.log(ratings);
